source/application/AMPIC887DriverServer/AMPIC887Controller.cpp

```diff
@@ -61,10 +114,954 @@ QString AMPIC887Controller::status() const
 		connectionStatus = QString("Connected with id %1").arg(id_);
 	}
 
-	return QString("%1\t\t\t%2\t\t\t%3")
+	return QString("%1\t\t\t%2\t\t\t%3\n%4")
 			.arg(name_)
 			.arg(hostname_)
-			.arg(connectionStatus);
+			.arg(connectionStatus)
+			.arg(controllerState_->statusString());
+}
+
+bool AMPIC887Controller::isBusy() const
+{
+	return isBusy_;
+}
+
+void AMPIC887Controller::clearErrorMessage()
+{
+	isBusy_ = true;
+
+	errorClearingTimer_.singleShot(30, this, SLOT(onErrorClearingTimerTimedOut()));
+}
+
+AMPIC887DataRecorderConfiguration AMPIC887Controller::recordConfig(int tableId) const
+{
+	return controllerState_->dataRecorderState()->recordConfig(tableId);
+}
+
+bool AMPIC887Controller::setRecordConfig(int tableId,
+										   const AMPIC887DataRecorderConfiguration &recorderConfig)
+{
+	QHash<int, AMPIC887DataRecorderConfiguration> singleValue;
+	singleValue.insert(tableId, recorderConfig);
+
+	return setRecordConfigs(singleValue);
+}
+
+QHash<int, AMPIC887DataRecorderConfiguration> AMPIC887Controller::recordConfigs() const
+{
+	return controllerState_->dataRecorderState()->recordConfigs();
+}
+
+bool AMPIC887Controller::setRecordConfigs(const QHash<int, AMPIC887DataRecorderConfiguration> recordConfigs)
+{
+	AMGCS2SetDataRecorderConfigurationCommand setRecordConfigCommand(recordConfigs);
+	runCommand(&setRecordConfigCommand);
+
+	if(setRecordConfigCommand.runningState() != AMGCS2Command::Succeeded) {
+		setError(setRecordConfigCommand.lastError());
+		return false;
+	}
+
+	controllerState_->dataRecorderState()->setRecordConfigs(recordConfigs);
+
+	return true;
+}
+
+QFlags<AMGCS2::AxisMovementStatus> AMPIC887Controller::movementStatus() const
+{
+	QFlags<AMGCS2::AxisMovementStatus> movements = QFlags<AMGCS2::AxisMovementStatus>();
+
+	if(xMotions_ > 0) {
+		movements |= AMGCS2::XAxisIsMoving;
+	}
+
+	if(yMotions_ > 0) {
+		movements |= AMGCS2::YAxisIsMoving;
+	}
+
+	if(zMotions_ > 0) {
+		movements |= AMGCS2::ZAxisIsMoving;
+	}
+
+	if(uMotions_ > 0) {
+		movements |= AMGCS2::UAxisIsMoving;
+	}
+
+	if(vMotions_ > 0) {
+		movements |= AMGCS2::VAxisIsMoving;
+	}
+
+	if(wMotions_ > 0) {
+		movements |= AMGCS2::WAxisIsMoving;
+	}
+
+	return movements;
+}
+
+bool AMPIC887Controller::isMoving() const
+{
+	return xMotions_ > 0 ||
```

Comment:
You could use `QFlags::testFlag()` here if you mapped these to full QFlags rather than just using an enum.